text/0002-modules.md

```diff
@@ -0,0 +1,57 @@
+are executed in the browser. Most major frameworks (React, Vue, Angular, etc) always use build-time modules, but do not
+always use in-browser modules. At OpenMRS, we use both.
+
+In-browser javascript modules are downloaded with one network request per module. Build-time modules are compiled away such
+that there is only one network request for a whole group (bundle) of build-time modules.
+
+[Browser support for modules](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#Browser_compatibility)
+started in Chrome 61, Firefox 60, Edge 16, and Safari 10.1. Internet Explorer does not support modules.
+
+A module loader is a javascript library that allows you to use modules in a browser that does not support them. This makes
+it possible for us to use modules in Internet Explorer.
+
+[SystemJS](https://github.com/systemjs/systemjs) is a minimalistic module loader that
+[polyfills](https://en.wikipedia.org/wiki/Polyfill_(programming)) modules so that you can use them in older browsers.
+When browsers fully support modules, the browser will be able to do what SystemJS is currently doing for us.
+
+## Reason for decision
+- Modules are a well-known programming language feature that allow for extensibility and separation of concerns.
+- Modules allow distribution implementors and maintainers to extend OpenMRS by using a browser spec, instead of something
+OpenMRS-specific. This means that extending OpenMRS is better understood and more approachable for the frontend community.
+- Modules create better [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns), since dependencies
+between modules are explicit and do not rely on global variables.
+- Modules can be lazy loaded, or not loaded at all. This allows OpenMRS distributions to choose which modules they wish
+to include in their code.
+
+## Alternatives
+Instead of using both in-browser modules and build-time modules, we could use build-time modules exclusively. This would
+simplify some things, but would mean that all frontend code would have to go through a single build process, which would be
+very difficult to get working when supporting multiple frameworks and codebases.
+
+Another alternative is to use neither in-browser modules nor build-time modules. This is how much Javascript code is still
+written, including all existing OpenMRS JSPs and GSPs. Not using modules makes it a lot more difficult to build an extensible frontend,
+since you ultimately must rely on load-order and global variables for code to interact with each other.
+Dependencies between global variables are not explicit and harder to manage.
+
+## Common practices (not enforced)
+- When implementing frontend code at OpenMRS, bundle many build-time modules into a single in-browser module by using a javascript bundler.
+- Create only one in-browser module for your application's components, instead of one in-browser module for each component.
```

Comment:
@joeldenning I have a few questions: 

Since the imports will be done at runtime, I assume end users can have the ability to add more than one in-browser module, will the app take a performance hit when this is done? Can we limit the number of in-browser modules? Is there a possibility that in-browser modules can conflict with each other?

Reply:
> Since the imports will be done at runtime, I assume end users can have the ability to add more than one in-browser module

Yep!

> will the app take a performance hit when this is done?

No. Adding a module means that the javascript code for the module will be downloaded and executed, the same as any other javascript file. But there's no performance overhead beyond just downloading and executing the javascript file.

> Can we limit the number of in-browser modules?

There's not a way of enforcing a limit to # of modules. But ultimately it's not so much about the number of modules as much as the number of bytes of javascript that are being downloaded and executed. 15 modules that are 1 kilobyte is wayyy better for performance than one module that is 1 module that is 15 megabytes. Think of a "module" as just a `<script>` element -- it's not so much how many you have as much as how many bytes they are.

> Is there a possibility that in-browser modules can conflict with each other?

The only unique thing about a module is its name. The code's imports, exports, and variables never conflict. In the following example, you could not add another `@ampath/appointments`, but you could replace it with a different one or create a new module that has some new functionality.

```json
{
  "imports": {
    "@ampath/appointments": "https://ampath.com/appointment.js"
  }
}
```